builtin/logical/pki/crl_util.go

```diff
@@ -199,29 +235,153 @@ func (cb *crlBuilder) requestRebuildIfActiveNode(b *backend) {
 
 	b.Logger().Info("Scheduling PKI CRL rebuild.")
 	// Set the flag to 1, we don't care if we aren't the ones that actually swap it to 1.
-	atomic.CompareAndSwapUint32(&cb.forceRebuild, 0, 1)
+	cb.forceRebuild.Store(true)
 }
 
 func (cb *crlBuilder) _doRebuild(ctx context.Context, b *backend, request *logical.Request, forceNew bool, ignoreForceFlag bool) error {
-	cb.m.Lock()
-	defer cb.m.Unlock()
+	cb._builder.Lock()
+	defer cb._builder.Unlock()
 	// Re-read the lock in case someone beat us to the punch between the previous load op.
-	forceBuildFlag := atomic.LoadUint32(&cb.forceRebuild)
-	if forceBuildFlag == 1 || ignoreForceFlag {
+	forceBuildFlag := cb.forceRebuild.Load()
+	if forceBuildFlag || ignoreForceFlag {
 		// Reset our original flag back to 0 before we start the rebuilding. This may lead to another round of
 		// CRL building, but we want to avoid the race condition caused by clearing the flag after we completed (An
 		// update/revocation occurred attempting to set the flag, after we listed the certs but before we wrote
 		// the CRL, so we missed the update and cleared the flag.)
-		atomic.CompareAndSwapUint32(&cb.forceRebuild, 1, 0)
+		cb.forceRebuild.Store(false)
 
 		// if forceRebuild was requested, that should force a complete rebuild even if requested not too by forceNew
-		myForceNew := forceBuildFlag == 1 || forceNew
+		myForceNew := forceBuildFlag || forceNew
 		return buildCRLs(ctx, b, request, myForceNew)
 	}
 
 	return nil
 }
 
+func (cb *crlBuilder) clearDeltaWAL(sc *storageContext) error {
+	// Clearing of the delta WAL occurs after a new complete CRL has been built.
+	walSerials, err := sc.Storage.List(sc.Context, deltaWALPath)
```

Comment:
I *think* we might have a race here as the thread that updates/adds new entries to the deltaWALPath (revoke api) does not grab the lock so we might delete a new entry if the timing is really bad?

Reply:
Ah, so we should grab the revoked storage lock? Is that safe to do? Should we have a closer scoped lock? Hmmm...

Reply:
I've included an alternate approach, let me know what you think.

Reply:
Yeah I like this approach 👍